### ISSUE TITLE
A new fix for LODES origin-destination matrices

### ISSUE THREAD

**Issue by santiagohermo** [diegogentilepassaro/min_wage_rent#101]
In this issue we will update the construction of the LODES OD matrices.

First, we will set the "destination threshold" to 1.

Second, we will fix the construction following these steps:
1. Aggregate from block-group level to tract level as now
2. Aggregate from tract level to ZCTA (instead of ZIP code)
3. Assign a ZIP code to each ZCTA based on the `geo_master`

Finally, as a sanity check, _assert_ that number of workers in home equals sum of number of workers in each workplace.

**_UPDATE MARCH 24_**

**GOAL**
Update geo-master construction so to have a zipcode-place-county level frame that can be then used as template for other datasets.

**STEPS**
1. Update geo-master: return a zipcode-place-county level dataset
2. update OD matrix construction 
    - Aggregate from block-group to tract as now
    - Aggregate from tract to zipcode (as now)
    - set destination threshold to 1 
3. Update MW construction, using new geo-master template

**Comment by gabrieleborg** [comment]
@diegogentilepassaro @santiagohermo I am working on this. But I keep forgetting what was the main issue here. Can you guys write it here? so we have it recorded :) thanks!

**Comment by santiagohermo** [comment]
> @diegogentilepassaro @santiagohermo I am working on this. But I keep forgetting what was the main issue here. Can you guys write it here? so we have it recorded :) thanks!

The main issue here is that, since in geo_master we keep one zip code per zcta, when we match the LODES to our crosswalk we lose ZIP codes (the ones we dropped in the crosswalk). Thus, we have missing MWs in the OD matrix, which means that some commuters are in practice being dropped out of the computations.

To avoid this, the construction of the OD matrix should be at the zcta level, and then we "replace" the ZCTA by our "chosen" zip code. That way we won't have any missing when matching the OD matrix to the MWs, since these are constructed on the geo_master. The assumption we will be making is that all the people in a zcta reside in the zip code that we select for that zcta.

Do you agree @diegogentilepassaro @gabrieleborg?

**Comment by diegogentilepassaro** [comment]
Agreed @santiagohermo! Thanks for writing it down very clearly 👍

**Comment by gabrieleborg** [comment]
thank you @santiagohermo !

**Comment by gabrieleborg** [comment]
Hey @santiagohermo @diegogentilepassaro! 

I am finishing implementing the changes discussed. But as I anticipated in the chat, a new issue with the geographies came up, and that made me (once again) start re-thinking about what are we trying to achieve with the geo-master approximations we are introducing. I think it is important to recap and reassess frequently on this because this is bringing a series of practical consequences we hadn't see in the first place, and so we need to carefully assess pros and cons of our choices. I know we talked extensively about this already, but the deeper we go in implementing the geo-master, the more carefully we need to evaluate the impact it has on our data. 
I will work backwards, by starting highlighting the problem that has appeared in this issue, and then try to summarize previous decisions so to provide a clearer summary: 
- We are trying to update the construction of the LODES OD matrix, which in turn we use to build our Exp. We want to **first** build a zcta-zcta matrix, and then use a 1:1 mapping between zipcode and zcta based on geo-master to finally have a zipcode-zipcode matrix.
    - PROBLEM: the geo-master assign each zipcode to a zcta-place-county combination, but this does not guarantee a 1:1 mapping between zipcode and zcta. In fact, there are multiple zipcodes that end up assigned to the same zcta. 
        - FIX: so far, the most basic thing to do would be to select, for each zcta, the zipcode associated to the zcta-zipcode pair with the most housing units as recorded in the geo-master. This however introduces **measurement error**: when we then build the experienced MW consider zipcode A: now we should have all destination zipcode info, but this will be an approximation of the true set of destination zipcodes, and this comes directly from the correction step because, for each zcta, we kept one zipcode only. So we are de-facto approximating a set of residence-workplace workflows with only the one whose destination zipcode had more houses in the associated zcta.  
- Previously we built it directly at the zipcode-zipcode level. 
    - PROBLEM:  some zipcodes didn't have a MW assigned. This, in turn, was caused by the fact that we decided to assign only one zipcode for each zcta-place-county. This also seems to introduce **measurement error**, but at this point maybe it's worth thinking which of the two is worse. 
        - WHY WE DID THAT: We originally decided to pick one zipcode for each zcta-place-county to avoid a m:m matching between zipcodes and zctas in the geo-master. We discussed this in #99 (see [here](https://github.com/diegogentilepassaro/min_wage_rent/pull/99#issuecomment-796011970)). There I don't remember exactly why we decided that, so you guys let me know if you recall it. 

Since we want to make the experienced MW an important part of the paper, I think we should really try to clearly compare the benefits and drawbacks of all these decisions we are taking. Is there a way to avoid such measurement error? you let me know what you think!

**Comment by santiagohermo** [comment]
Thanks @gabrieleborg! I thought that before we were selecting one zip code per zcta-place combination. Sad to hear we are not :(

I'd go with the first option. I think we are introducing measurement error in the geographical area of the zip code, but not in the commuting shares. Because the matrix is at the ZCTA level, and we have all ZCTAs in the US, we are counting every person in the construction of the matrix. We are just relabeling some ZCTAs to our most prefered ZIP code to add rents data. So the rents data is mismeasured since it doesn't quite represent the geographical area we are pasting it to.

I think we should be constructing everything MWs and LODES at the ZCTA level, and then we relabel those ZCTA to the ZIP code we like the most. (The entire problem would go away if Zillow gave us the listings, and we could aggregate them at the ZCTA label. May be one day.) Doing as before seems worse to me, since we were just throwing people out from the sample.

We should probably check, among the Zillow zip codes, how often they are different than ZCTAs, and report this in the paper.

What do you think?

**Comment by gabrieleborg** [comment]
Yeah it sounds good to me, thanks santi!

**Comment by diegogentilepassaro** [comment]
Thanks for flagging this @gabrieleborg!  I also thought that we were selecting one zip code per zcta-place combination :/

In principle I think I agree with @santiagohermo, but first I want to make sure I understand what the problem is. When you say:

PROBLEM: the geo-master assign each zipcode to a zcta-place-county combination, but this does not guarantee a 1:1 mapping between zipcode and zcta. In fact, there are multiple zipcodes that end up assigned to the same zcta.

Can you give a concrete example of how a 1:1 mapping from zipcode to zcta-place-county can produce multiple zipcodes that end up assigned to the same zcta? Also, do you know where in the code we are introducing this issue? 

We can workout this together if you want! Let me know and we can make a call tomorrow anytime :D

**Comment by gabrieleborg** [comment]
Hey guys! @diegogentilepassaro @santiagohermo 

Thanks for your comments, they allowed me to think a bit more about the problem, and to clarify a bit better. First, here is an example of what I mean when I say that the geo-master assign more than one zip code to a zcta: 
<img width="1018" alt="Screen Shot 2021-03-23 at 10 25 01 AM" src="https://user-images.githubusercontent.com/30894427/112166772-51608e80-8bc6-11eb-8eb6-6b1e66cf5f33.png">
Here you can see how, even if we are selecting one zip code per zcta-place combination (IF we don't consider rural places coded as 99999), we still have that zcta 1230 has 2 different zip codes assigned: 1230 and 1252. The problem with the OD matrix is that we don't have info on all the zcta-place combinations, but we only have zcta. This is a byproduct of our construction steps (i.e. we aggregate blocks to tracts, and then crosswalk tracts to zcta). This happens in `base/lodes/code/make_odmatrix.R`. A potential solution could be to change that first step and assign each block group to a proper zcta-place combination. BUT this requires a new crosswalk that we don't have, and the infinite loop of geographies starts again. 

Also, let me clarify why I think there could be measurement error in the commuting shares. I'll try to do that by writing down the logic behind the construction of our OD matrix: 

*TRACT TO ZIP CODE (OLD)*
- For each tract, sum workers in each destination zipcode (with weights)
- For each destination zipcode, sum residents in each origin zipcode (with weights)

*TRACT TO ZCTA TO ZIPCODE*
- For each tract, sum workers in each destination zcta(with weights)
- For each destination zcta, sum residents in each origin zcta(with weights)
- But now, there are more than one zipcodes in certain zctas
- For each home zcta, I assign to each destination zcta one zipcode. This is the zipcode with the most residential units within the zcta, but It may not be the only one present. If it's not, I am therefore saying that from a given home zcta, all workers that work in each of the zipcodes within the target zcta, actually work in only 1 zipcode. 
- Then, for each destination zipcode mapped, I take home zctas and assign them to 1 zipcode only (the one with the most houses). So once again, for any destination zipcode (already potentially overestimating the number of workers there) I am assigning the residence of all workers living in potentially multiple zipcodes within a given home zcta to just the one with the most houses.

**Comment by gabrieleborg** [comment]
As per the last meeting, we are updating the **GOAL** of this issue so to have a comprehensive geo-master that allows to build a MW dataset for the whole US (as opposed to Zillow Zip Codes only). See Details in updated intro. After this, we whould be able to obtain a LODES OD matrix directly at the zip code level with no missing MW info for destination zip codes.

**Comment by santiagohermo** [comment]
I just implemented the agreed-upon changes to the `geo_master` @diegogentilepassaro @gabrieleborg. I'm keeping zipcode-county-place combinations based on the rule described [here](https://github.com/diegogentilepassaro/min_wage_rent/pull/99#issuecomment-796011970). Can you check that the changes make sense to you?

I'll re-run OD matrix, following steps [here](https://github.com/diegogentilepassaro/min_wage_rent/issues/101#issue-834162348). I think that we can leave the re-run of MW variables to #102, since @diegogentilepassaro has already changed those scripts a bit.

**Comment by gabrieleborg** [comment]
Thanks @santiagohermo ! I was about to push the same thing to! I'll check and post my comment soon

**Comment by santiagohermo** [comment]
Sorry @gabrieleborg! I thought you were already on a plane haha.

**Comment by santiagohermo** [comment]
Once you give me the OK on the `geo_master` I'll run the OD matrix on the server @gabrieleborg! Let me know.

**Comment by gabrieleborg** [comment]
I reviewed (once again ahah) the geo-master construction and tried to keep track of all zipcode issues we encounter. In general, I think the changes Santi did are great. I just left an inline comment in the commit, and in my modification I just wanted to add a dummy variable to select, for each zipcode-county-place combination, only the zip code with the most number of houses. What do you think? I know we'll most likely take weighted averages, but I remember we discussed a bit about possible alternatives, and this could be an easy way to subsequently filter the geo master bringing it from the zipcode-place-county level to the zip code level. 

Here are some info and more detailed comment:
1. According to our zcta-zip 2019 crosswalk, there were 41107 zip code assigned to 33166 zcta (lower because of PO boxes, universities, etc)
2. We then do an m:m merge with the zcta-place-county-state master. We obtain 40488 unique zipcodes (we lose some as few don't have zcta assigned, and some other reason I haven't figured out clearly. I suspect it is because the master crosswalk we downloaded does not have puerto rico zipcodes, and maybe military ones and the zip-zcta crosswalk does?)
    - Of those 40488, 3316 are in the Zillow sample
        - Of those 3316, 1758 already have a unique zipcode-place-county combination, while 1558 have more than one. 
3. How can we get from a zipcode-place-county crosswalk to a zip code level one (e.g. when we have to build MW, or lodes OD matrix)? 
    - The most immediate way is to use the housing weights we have. Theoretically, there could be an issue here (but hey, we don't have it in practice!) This is because weights are at the zcta-place-county level. So what shall we do when we have, say, 2 zip codes within the same zcta-place-county combination? In our dataset, they are assigned the same weight. **BUT**, I checked and it turns out the following: Of those 1558 zip codes where we have more than one zipcode-place-county combination (and that would therefore require weights) NONE belongs to a zcta-place-county combination with another different zipcode. The two sets of zipcodes (those with more zipcode-place-county combinations and those that share a zcta-place-county combination) are not overlapping. So we are good, and the resulting MW associated to each zip code will be correctly computed as a weighted mean.     
    - Another possibility would be to select one zipcode-place-county combination for those 1558 zipcodes that have more than one. One natural way is to choose the one with the highest number of houses. Once again, even if weights are at the zcta-place-county level this is not a problem for the aforementioned reason. I created a dummy variable already selecting such subsample. What do you want to do about it @santiagohermo @diegogentilepassaro ? I think I already commited it, so if we decinde not to keep it maybe delete those lines before running LODES OD!
    
Re-running Lodes here and MW in derived sounds pefect to me. Let me know how it goes!

**Comment by gabrieleborg** [comment]
Another thing to add. since now we are outputting a zipcode-place-county level geo-master, I don't think we need all those rules to select zipcodes within zcta anymore ( those in the Zillow sample, those urban etc). The reason is that now we have all the zipcodes, and we don't have to sacrifice any. The only choice we have to make is how to go from zipcode-place-county to zipcode level, but this doen't imply dropping any, only pick a zipcode-place-county combination for each zipcode.

**Comment by santiagohermo** [comment]
Thanks @gabrieleborg! Replies:

- I replied to the inline comment [here](https://github.com/diegogentilepassaro/min_wage_rent/commit/c884b1387566ced8acb7c1f433ce524218c7ad9a#r48746892). Just to reiterate, do we or do we not want the geo_master to be keyed by zipcode-county-place? If yes, I think we need to do some dropping, don't we?
  To extend a bit. The geo_master had multiple zcta-zipcode-county-place combinations. The filtering I implemented before tried to drop the zcta part, and keep zipcode-county-place combinations only. Then, on top of that, we can create a dummy to select one zipcode per county-place, as you say. Do you agree?

> * Of those 40488, 3316 are in the Zillow sample
>   * Of those 3316, 1758 already have a unique zipcode-place-county combination, while 1558 have more than one.

- Thanks for the breakdown! Very useful. So whether we take weighted averages or max may change things a bit, since for nearly half of the zip codes we'll be doing things differently.

- About your point 3. That's great, thanks for checking! If we ever decide to make an R/Stata package out of this, we'll need to get weights at the zip code level though haha. It would be nice to solve this in general, but since weights that vary at zip code level are hard to get, and we don't need them for this project as you say, I think we are good.

**Comment by santiagohermo** [comment]
I implemented a few changes that should implement this:

> * To extend a bit. The geo_master had multiple zcta-zipcode-county-place combinations. The filtering I implemented before tried to drop the zcta part, and keep zipcode-county-place combinations only. Then, on top of that, we can create a dummy to select one zipcode per county-place, as you say. Do you agree?

Let me know if you disagree @gabrieleborg! I'll set things running

**Comment by gabrieleborg** [comment]
I just checked: you can type `isid zipcode place_code county` right after we do the m:m merge between the zip-zcta and the master crosswalk, and you'll see that combination is already unique. That is the level for the geo-master we want, since it is the level to then build MW. now we don't really care about zcta anymore so we don't need to drop anything. The only problem could have been the weighting issue, but that we saw it's fine. This to say that I don't think we need lines 61-68. 

I already created the dummy choosing one zipcode per zipcode-place-county combination, you guys just choose whether you want to keep it

**Comment by santiagohermo** [comment]
Thanks @gabrieleborg! [That](https://github.com/diegogentilepassaro/min_wage_rent/issues/101#issuecomment-808272830) makes perfect sense to me. I'll replace those lines by the `isid` statement once the `lodes` step is done.

**Comment by santiagohermo** [comment]
About [this](https://github.com/diegogentilepassaro/min_wage_rent/issues/101#issuecomment-808272830) @gabrieleborg. In https://github.com/diegogentilepassaro/min_wage_rent/commit/5ee87398c9cf57b7c2f1d5f70a7551ac5ef3671e I added that `isid` statement and deleted several lines, and the result is identical. Nice catch!!

---

About LODES @diegogentilepassaro @gabrieleborg. In completed a run [here](https://github.com/diegogentilepassaro/min_wage_rent/commit/5ee87398c9cf57b7c2f1d5f70a7551ac5ef3671e) and added output to google drive.  

Results changed a lot! As can be seen in [the data manifest](https://github.com/diegogentilepassaro/min_wage_rent/commit/e269a8fd379f95ba2cce3207db2c775eedd7e441#diff-c6a90fb5e4fde7929fddbc2bcb06b6db636987e18029c37c279198267152a309), the number of residence-workplace combinates increased between 10 and 15 times depending on the state. Crazy stuff.

Since we already reviewed all the steps here, I'll quickly merge this to #85 and update the branch for #102.

**Comment by santiagohermo** [comment]
Continues in #103
